### PR TITLE
[Behat] Implementation Improvements of the `ResponseCheckerInterface`

### PR DESCRIPTION
**Pull request by Rafikooo**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13                   |
| Bug fix?        | no                                                      |
| New feature?    | no                                                      |
| BC breaks?      | no                                                      |
| Deprecations?   | no |
| License         | MIT                                                          |

### CODE REVIEW

**Review by Wojdylak — DISMISSED** [review]

File: src/Sylius/Behat/Context/Setup/ChannelContext.php [hunk 1]
```diff
@@ -182,7 +183,7 @@ public function itsDefaultTaxRateIs(ChannelInterface $channel, ZoneInterface $de
      * @Given /^(this channel) has contact email set as "([^"]+)"$/
      * @Given /^(this channel) has no contact email set$/
      */
-    public function thisChannelHasContactEmailSetAs(ChannelInterface $channel, $contactEmail = null)
+    public function thisChannelHasContactEmailSetAs(ChannelInterface $channel, $contactEmail = null): void
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Context/Setup/ChannelContext.php]:
```suggestion
    public function thisChannelHasContactEmailSetAs(ChannelInterface $channel, ?string $contactEmail = null): void
```

File: src/Sylius/Behat/Client/ResponseChecker.php [hunk 2]
```diff
@@ -132,18 +132,54 @@ public function hasItemWithValue(Response $response, string $key, $value): bool
...
-    /** @param string|int $value */
-    public function hasSubResourceWithValue(Response $response, string $subResource, string $key, $value): bool
-    {
+    public function hasValueInAnySubresourceObjectCollection(
+        Response $response,
+        string $subResource,
+        string $key,
+        int|string $expectedValue,
+    ): bool {
         foreach ($this->getResponseContentValue($response, $subResource) as $resource) {
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Client/ResponseChecker.php]:
```suggestion
        $resources = $this->getResponseContentValue($response, $subResource);
        foreach ($resources as $resource) {
```
To ease debugging a bit

File: src/Sylius/Behat/Client/ResponseChecker.php [hunk 3]
```diff
@@ -132,18 +132,54 @@ public function hasItemWithValue(Response $response, string $key, $value): bool
...
 
-    /** @param string|int $value */
-    public function hasSubResourceWithValue(Response $response, string $subResource, string $key, $value): bool
-    {
+    public function hasValueInAnySubresourceObjectCollection(
+        Response $response,
+        string $subResource,
+        string $key,
+        int|string $expectedValue,
+    ): bool {
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Client/ResponseChecker.php]:
```suggestion
    public function hasSubresourceWithValuesInCollection(
        Response $response,
        string $subResource,
        /** @var array<string, string|int> $expectedValues */
        array $expectedValues,
    ): bool {
```
Would be nice to be able to pass an array of parameters that are expected, so we can check whether a singular object in the collection has a specific id/key/position/whatever at once.

Comment (Rafikooo) [reply src/Sylius/Behat/Client/ResponseChecker.php]: I've implemented the idea, please take a look if it's what you had in mind

File: src/Sylius/Behat/Client/ResponseChecker.php [hunk 4]
```diff
@@ -132,18 +132,51 @@ public function hasItemWithValue(Response $response, string $key, $value): bool
...
+        string $subResourceName,
+        array $expectedValues,
+    ): bool {
+        $resourceCollection = $this->getResponseContentValue($response, $subResourceName);
+
+        $this->assertIsArray($resourceCollection);
+
+        foreach ($resourceCollection as $resource) {
+            $this->assertIsArray($resource);
+            $this->assertAllExpectedKeysArePresent($expectedValues, $resource);
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Client/ResponseChecker.php]: Won't this potentially fail if a property is nullable? I don't really see a point to it.

File: src/Sylius/Behat/Client/ResponseChecker.php [hunk 5]
```diff
@@ -132,18 +132,50 @@ public function hasItemWithValue(Response $response, string $key, $value): bool
...
 
-    /** @param string|int $value */
-    public function hasSubResourceWithValue(Response $response, string $subResource, string $key, $value): bool
-    {
-        foreach ($this->getResponseContentValue($response, $subResource) as $resource) {
-            if ($resource[$key] === $value) {
-                return true;
+    public function hasValuesInAnySubresourceObjectCollection(
+        Response $response,
+        string $subResourceName,
```
Comment (NoResponseMate) [inline src/Sylius/Behat/Client/ResponseChecker.php]:
```suggestion
        string $subResource,
```
To be consistent

**Review by NoResponseMate — DISMISSED** [review]

**Review by NoResponseMate — APPROVED** [review]

**Review by jakubtobiasz — APPROVED** [review]

**Comment by jakubtobiasz** [comment]
Thank you, @Rafikooo!

**Pull request merged**